lib/msal-browser/src/broker/BrokerClient.ts

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { BrokerOptions } from "../config/Configuration";
+import { Logger, AuthenticationResult, ClientAuthError } from "@azure/msal-common";
+import { BrokerHandshakeRequest } from "./BrokerHandshakeRequest";
+import { BrokerHandshakeResponse } from "./BrokerHandshakeResponse";
+
+/**
+ * Embedded application in a broker scenario.
+ */
+export class BrokerClient {
+    private brokerOptions: BrokerOptions;
+    private brokeringEnabled: boolean;
+    private logger: Logger;
+    private clientId: string;
+    private version: string;
+
+    constructor(brokerOptions: BrokerOptions, logger: Logger, clientId: string, version: string) {
+        this.brokerOptions = brokerOptions;
+        this.logger = logger;
+        this.clientId = clientId;
+        this.version = version;
+
+        if (!this.brokerOptions.trustedBrokerDomains || this.brokerOptions.trustedBrokerDomains.length < 1) {
+            this.logger.info("Application was identified as an embedded app in a broker scenario, but no trusted broker domains were provided.");
+            this.brokeringEnabled = false;
+        } else {
+            this.brokeringEnabled = true;
+        }
+    }
+
+    async initiateHandshake(): Promise<void> {
+        if (!this.brokeringEnabled) {
+            this.logger.verbose("Brokering is not enabled, handshake was not performed.");
+            return;
+        }
+
+        try {
+            await this.sendHandshakeRequest();
+        } catch (e) {
+            this.logger.error(e);
+            this.brokeringEnabled = false;
+            throw e;
+        }
+    }
+
+    private async sendHandshakeRequest(): Promise<BrokerHandshakeResponse> {
+        return new Promise<BrokerHandshakeResponse>((resolve: any, reject: any) => {
+            const timeoutId = setTimeout(() => {
+                this.logger.warning("Broker handshake timed out");
+                reject(new ClientAuthError("message_broker_timeout", "Message broker timed out"));
```

Comment:
It may be a good idea to replace this `ClientAuthError` instantiation with a static function for consistency (i.e. ClientAuthError.createMessageBrokerTimeoutError()).